axum/src/extract/state.rs

```diff
@@ -295,6 +295,49 @@ use std::{
 ///
 /// In general however we recommend you implement `Clone` for all your state types to avoid
 /// potential type errors.
+///
+/// # Shared mutable state
+///
+/// [As state is global within a `Router`][global] you can't directly get a mutable reference to
+/// the state.
+///
+/// The most basic solution is to use an `Arc<Mutex<_>>`. Which kind of mutex you need depends on
+/// your use case. See [the tokio docs] for more details.
+///
+/// Note that holding a locked `std::sync::Mutex` across `.await` points will result in `!Send`
+/// futures which are incompatible with axum. If you need to hold a mutex across `.await` points
+/// consider using a `tokio::sync::Mutex` instead.
+///
+/// ## Example
+///
+/// ```
+/// use axum::{Router, routing::get, extract::State};
+/// use std::sync::{Arc, Mutex};
+///
+/// #[derive(Clone)]
+/// struct AppState {
+///     data: Arc<Mutex<String>>,
+/// }
```

Comment:
I'm a little worried that people will do fine-grained locking where more coarse locking is the right solution, based on this example. I'm not sure just showing an `Arc<Mutex<_>>` with everything inside locked at once is better either though...

Reply:
@jplatte  Maybe something like the following can be added, in order to make your point super explicit?
```
#[derive(Clone)]
struct AppState {
      // This is just an example usage, not a best practice.
      // Choose a setup that matches you use specific use case.
      data: Arc<Mutex<String>>,
}
```